src/libponyc/expr/array.c

```diff
@@ -10,14 +10,259 @@
+  ast_setdata(array_type, ast_get(scope, stringtab("Array"), NULL));
+
+  return array_type;
+}
+
+static void find_possible_element_types(pass_opt_t* opt, ast_t* ast,
+  astlist_t** list)
+{
+  switch(ast_id(ast))
+  {
+    case TK_NOMINAL:
+    {
+      AST_GET_CHILDREN(ast, package, name, typeargs, cap, eph);
+
+      // If it's an actual Array type, note it as a possibility and move on.
+      if(stringtab("Array") == ast_name(name))
+      {
+        *list = astlist_push(*list, ast_child(typeargs));
+        return;
+      }
+
+      // Otherwise, an Array-matching type must be an interface.
+      ast_t* def = (ast_t*)ast_data(ast);
+      if((def == NULL) || (ast_id(def) != TK_INTERFACE))
+        return;
+
+      // The interface must have an apply method for us to find it.
+      ast_t* apply = ast_get(def, stringtab("apply"), NULL);
+      if((apply == NULL) || (ast_id(apply) != TK_FUN))
+        return;
+
+      // The apply method must match the signature we're expecting.
+      AST_GET_CHILDREN(apply, receiver_cap, apply_name, type_params, params,
+        ret_type, question);
+      if((ast_id(receiver_cap) != TK_BOX) ||
+        (ast_id(type_params) != TK_NONE) ||
+        (ast_childcount(params) != 1) ||
+        (ast_id(question) != TK_QUESTION))
+        return;
+
+      ast_t* param = ast_child(params);
+      ast_t* param_type = ast_childidx(param, 1);
+      if(ast_name(ast_childidx(param_type, 1)) != stringtab("USize"))
+        return;
+
+      // Based on the apply method we try to figure out the element type.
+      ast_t* elem_type = ret_type;
+
+      ast_t* typeparams = ast_childidx(def, 1);
+      if(ast_id(typeparams) == TK_TYPEPARAMS)
+        elem_type = reify(ret_type, typeparams, typeargs, opt, true);
+
+      if((ast_id(elem_type) == TK_ARROW) &&
+        (ast_id(ast_child(elem_type)) == TK_THISTYPE))
+        elem_type = ast_childidx(elem_type, 1);
+
+      // Construct a guess of the corresponding Array type.
+      // Use iso^ so that the object cap isn't a concern for subtype checking.
+      ast_t* array_type = build_array_type(ast, elem_type, TK_ISO);
+
+      // The guess type must be a subtype of the interface type.
+      if(!is_subtype(array_type, ast, NULL, opt))
+      {
+        ast_free_unattached(array_type);
+        return;
+      }
+
+      ast_free_unattached(array_type);
+
+      // Note this as a possible element type and move on.
+      *list = astlist_push(*list, elem_type);
+    }
+
+    case TK_ARROW:
+      find_possible_element_types(opt, ast_childidx(ast, 1), list);
+      return;
+
+    case TK_TYPEPARAMREF:
+    {
+      ast_t* def = (ast_t*)ast_data(ast);
+      pony_assert(ast_id(def) == TK_TYPEPARAM);
+      find_possible_element_types(opt, ast_childidx(def, 1), list);
+      return;
+    }
+
+    case TK_UNIONTYPE:
+    case TK_ISECTTYPE:
+    {
+      for(ast_t* c = ast_child(ast); c != NULL; c = ast_sibling(c))
+        find_possible_element_types(opt, c, list);
+    }
```

Comment:
I'd feel more comfortable with a `break` or `return` here, to avoid problems in future refactorings.

Reply:
Good call. Fixed in 943c1ba.